File: ios/Stopwatch/domain/utils/impl/CalculateLapsStatusesImpl.swift
class CalculateLapsStatusesImpl: CalculateLapsStatuses {
    func execute(_ laps: [Lap]) -> [Lap] {
        if laps.count < 2 {
            return laps.map {
                Lap(
                    index: $0.index,
                    milliseconds: $0.milliseconds,
                    status: .done
                )
            }
        }
        
        var best = laps[0]
        var worst = laps[0]
        
        for lap in laps {
            if lap.milliseconds < best.milliseconds {
                best = lap
            } else if lap.milliseconds > worst.milliseconds {
                worst = lap
            }
        }
        
        return laps.map {
            let status: LapStatus = switch($0.index) {
            case best.index: .best
            case worst.index: .worst
            default: .done
            }
            
            return Lap(
                index: $0.index,
                milliseconds: $0.milliseconds,
                status: status
            )
        }
    }
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/domain/utils/impl/CalculateLapsStatusesImpl.kt
package com.rafaeltmbr.stopwatch.domain.utils.impl

import com.rafaeltmbr.stopwatch.domain.entities.Lap
import com.rafaeltmbr.stopwatch.domain.utils.CalculateLapsStatuses

class CalculateLapsStatusesImpl : CalculateLapsStatuses {
    override fun execute(laps: List<Lap>): List<Lap> {
        if (laps.size < 3) {
            return laps.map {
                it.copy(
                    status = if (it == laps.last()) Lap.Status.CURRENT else Lap.Status.DONE
                )
            }
        }

        val current = laps.last()
        var best = laps[0]
        var worst = laps[0]

        for (lap in laps.subList(0, laps.size - 1)) {
            if (lap.milliseconds < best.milliseconds) {
                best = lap
            } else if (lap.milliseconds > worst.milliseconds && laps.size > 2) {
                worst = lap
            }
        }

        return laps.map {
            Lap(
                index = it.index,
                milliseconds = it.milliseconds,
                status = when (it) {
                    current -> Lap.Status.CURRENT
                    best -> Lap.Status.BEST
                    worst -> Lap.Status.WORST
                    else -> Lap.Status.DONE
                }
            )
        }
    }
}